cmd/bucket-metadata-sys.go

```diff
@@ -143,62 +142,126 @@ func (sys *BucketMetadataSys) Get(bucket string) (BucketMetadata, error) {
+	}
+	return meta.lifecycleConfig, nil
+}
+
+// GetNotificationConfig returns configured notification config
+func (sys *BucketMetadataSys) GetNotificationConfig(bucket string) (*event.Config, error) {
+	meta, err := sys.GetConfig(bucket)
+	if err != nil {
+		return nil, err
+	}
+	return meta.notificationConfig, nil
+}
+
+// GetSSEConfig returns configured SSE config
+func (sys *BucketMetadataSys) GetSSEConfig(bucket string) (*bucketsse.BucketSSEConfig, error) {
+	meta, err := sys.GetConfig(bucket)
+	if err != nil {
+		if errors.Is(err, errConfigNotFound) {
+			return nil, BucketSSEConfigNotFound{Bucket: bucket}
 		}
-		var err error
-		meta, err = loadBucketMetadata(GlobalContext, objAPI, bucket)
-		if err != nil {
-			return nil, err
+		return nil, err
+	}
+	if meta.sseConfig == nil {
+		return nil, BucketSSEConfigNotFound{Bucket: bucket}
+	}
+	return meta.sseConfig, nil
+}
+
+// GetPolicyConfig returns configured bucket policy
+func (sys *BucketMetadataSys) GetPolicyConfig(bucket string) (*policy.Policy, error) {
+	meta, err := sys.GetConfig(bucket)
+	if err != nil {
+		if errors.Is(err, errConfigNotFound) {
+			return nil, BucketPolicyNotFound{Bucket: bucket}
 		}
-		sys.metadataMap[bucket] = meta
+		return nil, err
+	}
+	if meta.policyConfig == nil {
+		return nil, BucketPolicyNotFound{Bucket: bucket}
 	}
+	return meta.policyConfig, nil
+}
 
-	var configData []byte
-	switch configFile {
-	case bucketPolicyConfig:
-		configData = meta.PolicyConfigJSON
-	case bucketNotificationConfig:
-		configData = meta.NotificationXML
-	case bucketLifecycleConfig:
-		configData = meta.LifecycleConfigXML
-	case bucketSSEConfig:
-		configData = meta.EncryptionConfigXML
-	case bucketTaggingConfigFile:
-		configData = meta.TaggingConfigXML
-	case objectLockConfig:
-		configData = meta.ObjectLockConfigurationXML
-	case bucketQuotaConfigFile:
-		configData = meta.QuotaConfigJSON
-	default:
-		return nil, fmt.Errorf("Unknown bucket %s metadata update requested %s", bucket, configFile)
+// GetQuotaConfig returns configured bucket quota
+func (sys *BucketMetadataSys) GetQuotaConfig(bucket string) (*madmin.BucketQuota, error) {
+	meta, err := sys.GetConfig(bucket)
+	if err != nil {
+		return nil, err
 	}
+	return meta.quotaConfig, nil
+}
 
-	if len(configData) == 0 {
-		return nil, errConfigNotFound
+// GetConfig returns a specific configuration from the bucket metadata.
+func (sys *BucketMetadataSys) GetConfig(bucket string) (BucketMetadata, error) {
+	objAPI := newObjectLayerWithoutSafeModeFn()
+	if objAPI == nil {
+		return newBucketMetadata(bucket), errServerNotInitialized
+	}
+
+	if globalIsGateway {
+		return newBucketMetadata(bucket), NotImplemented{}
+	}
+
+	if bucket == minioMetaBucket {
+		return newBucketMetadata(bucket), errInvalidArgument
 	}
 
-	return configData, nil
+	sys.Lock()
+	defer sys.Unlock()
+	meta, ok := sys.metadataMap[bucket]
+	if ok {
+		return meta, nil
+	}
+	meta, err := loadBucketMetadata(GlobalContext, objAPI, bucket)
```

Comment:
The reason for the fallback was to avoid this very costly operation when not found. This is especially bad since you are holding the lock.

What is the reason for removing it?

Reply:
It is a safety for an entry, I am not a fan either but it is just there is no good solution for missed updates. Rather than continuously running a thread which reloads inconsistent metadata from drives.

Reply:
We expect this to be rare anyways because this only happens when the entry is missing and the first call we just reload to memory.